### PR TITLE
Simulate faults and recoveries

### PR DESCRIPTION
**Pull request by acruikshank**
### Motivation

To better simulate all actor processes we need to simulate miner behavior when storage faults. This PR adds a mechanism to simulate and track faults which either result in declared fault messages or sectors being skipped when PoSts are submitted. Similarly, miners can recover their faults, so the simulation can now add a probability that a fault will be recovered resulting in a declare fault recovered message.

This PR also:

1. Introduces a `math.go` file with a new `RateIterator` to contain some of the stochastic machinery used by the simulation.
2. Introduces the `RateIterator` struct that will call a function some number of times each tick based on a rate and the Poisson distribution.

### CODE REVIEW

File: support/agent/miner_agent.go [hunk 1]
```diff
@@ -156,34 +195,187 @@ func (ma *MinerAgent) Tick(v SimState) ([]message, error) {
...
+		return nil, errors.Errorf("could not check if %d in deadline %d partition %d is faulty",
+			recoveryNumber, recoveryDlInfo.Index, pIdx)
+	} else if !set {
+		return nil, errors.Errorf("recovery %d in deadline %d partition %d was not a fault",
+			recoveryNumber, recoveryDlInfo.Index, pIdx)
+	}
+
+	// If it's too late, schedule recovery rather than declaring it
+	if recoveryDlInfo.FaultCutoffPassed() {
+		ma.operationSchedule.ScheduleOp(recoveryDlInfo.Close, recoverSectorAction{
```
Comment (ZenGround0) [inline support/agent/miner_agent.go:342]: Out of curiosity do we go through the operation schedule when we terminate or expire a sector to remove all outstanding operations on terminated sectors?

**Review by ZenGround0 — APPROVED** [review]
I'm missing a fair bit of context from not reading through the whole framework last review but the changes here look good.

File: support/agent/miner_agent.go (hunk 1 above)
Comment (acruikshank) [reply support/agent/miner_agent.go:354]: we do now.

File: support/agent/math.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+import (
+	"math"
+	big2 "math/big"
+	"math/rand"
+
+	"github.com/filecoin-project/go-state-types/abi"
+)
+
+type RateIterator struct {
```
Comment (anorth) [inline support/agent/math.go:11]: Brief documentation please

File: support/agent/miner_agent.go [hunk 3]
```diff
@@ -115,75 +134,260 @@ func NewMinerAgent(owner address.Address, worker address.Address, idAddress addr
...
+		}
+		messages = append(messages, msgs...)
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	// Recover sectors.
+	// Rate must be multiplied by the number of live sectors
+	recoveryRate := ma.Config.RecoveryRate * float64(len(ma.faultySectors))
```
Comment (anorth) [inline support/agent/miner_agent.go:212]: Comment disagrees with code

File: support/agent/miner_agent.go [hunk 4]
```diff
@@ -253,46 +471,149 @@ func (ma *MinerAgent) scheduleNextProof(v SimState, dlIdx uint64) error {
...
-// create precommit message and activation trigger
-func (ma *MinerAgent) createPrecommit(currentEpoch abi.ChainEpoch, sectorNumber abi.SectorNumber, sectorActivation abi.ChainEpoch) message {
-	params := miner.PreCommitSectorParams{
-		SealProof:     ma.Config.ProofType,
-		SectorNumber:  sectorNumber,
-		SealedCID:     sectorSealCID(ma.rnd),
-		SealRandEpoch: currentEpoch - 1,
-		Expiration:    ma.sectorExpiration(currentEpoch),
+func filterSlice(ns []uint64, toRemove map[uint64]bool) []uint64 {
+	var nextLive []uint64
```
Comment (anorth) [inline support/agent/miner_agent.go:574]: Consider pre-allocating, maybe to `len(ns) - len(toRemove)`

File: support/agent/miner_agent.go [hunk 5]
```diff
@@ -340,6 +656,45 @@ func sectorSealCID(rnd *rand.Rand) cid.Cid {
...
+	var err error
+	part.sectors, err = bitfield.SubtractBitField(part.sectors, newExpired)
+	if err != nil {
+		return err
+	}
+	part.faults, err = bitfield.SubtractBitField(part.faults, newExpired)
+	if err != nil {
+		return err
+	}
+	part.faults, err = bitfield.SubtractBitField(part.faults, newExpired)
```
Comment (anorth) [inline support/agent/miner_agent.go:683]: This is duplicated 4 lines above

**Review by anorth — APPROVED** [review]

File: support/agent/miner_agent.go (hunk 1 above)
Comment (acruikshank) [reply support/agent/miner_agent.go:354]: We don't, but only `registerSector` and `delayedRecoveryMessage` operate on individual on-chain sectors and they both handle missing sectors gracefully.

**Pull request merged**